pkg/kubelet/container/interface.go

```diff
@@ -0,0 +1,338 @@
+	// Network is the path to the network namespace.
+	Network string
+}
+
+// LinuxSandBoxStatus contains status specific to Linux sandboxes.
+type LinuxPodSandboxStatus struct {
+	// Namespaces contains paths to the sandbox's namespaces.
+	Namespaces *Namespaces
+}
+
+// PodSandboxResources contains the CPU/memory resource requirements.
+type PodSandboxResources struct {
+	// CPU resource requirement.
+	CPU api.ResourceRequirements
+	// Memory resource requirement.
+	Memory api.ResourceRequirements
+}
+
+// This is to distinguish with existing ContainerID type, which includes a
+// runtime type prefix (e.g., docker://). We may rename this later.
+type RawContainerID string
+
+// ContainerRuntime provides methods for container lifecycle operations, as
+// well as listing or inspecting existing containers.
+type ContainerRuntime interface {
+	// Create creates a container in the sandbox, and returns the ID
+	// of the created container.
+	Create(config *ContainerConfig, sandboxConfig *PodSandboxConfig, sandboxID PodSandboxID) (RawContainerID, error)
+	// Start starts a created container.
+	Start(RawContainerID ContainerID) error
+	// Stop stops a running container with a grace period (i.e., timeout).
+	Stop(RawContainerID string, timeout int) error
+	// Remove removes the container.
+	Remove(RawContainerID string) error
+	// List lists the existing containers that match the ContainerFilter.
+	// The returned list should only include containers previously created
+	// by this ContainerManager.
+	List(filter ContainerFilter) ([]Container, error)
+	// Status returns the status of the container.
+	Status(RawContainerID string) (ContainerStatus, error)
+}
+
+// ContainerCommandExecutor provides methods to run a command in the container.
+// TODO: Should we merge this with ContainerRuntime?
+type ContainerCommandExecutor interface {
+	// Exec executes a command in the container.
+	Exec(RawContainerID string, cmd []string, streamOpts StreamOptions) error
+}
+
+type ContainerConfig struct {
+	// Name of the container.
+	Name string
+	// Image to use.
+	Image ImageSpec
+	// Command to execute (i.e., entrypoint for docker)
+	Command []string
+	// Args for the Command (i.e., command for docker)
+	Args []string
+	// Current working directory of the command.
+	WorkingDir string
+	// List of environment variable to set in the container
+	Env []KeyValue
+	// Mounts specifies mounts for the container
+	Mounts []Mount
+	// Labels are key value pairs that may be used to scope and select individual resources.
+	Labels Labels
+	// Annotations is an unstructured key value map that may be set by external
+	// tools to store and retrieve arbitrary metadata.
+	Annotations map[string]string
+	// Privileged runs the container in the privileged mode.
+	Privileged bool
+	// ReadOnlyRootFS sets the root filesystem of the container to be
+	// read-only.
+	ReadOnlyRootFS bool
+	// Path to store the container log on the host (i.e., outside of the
+	// sandbox).
+	LogPath string
```

Comment:
> I'm quite confused by this discussion now.

Me too. There hasn't been any disagreement with the need for better isolation and yet we are continuing to talk about `journald` which clearly isn't satisfying all the requirements for kubernetes. If there is an alternative solution that provides all the isolation requirements I mentioned, I'm all ears. My goal is to get most kubernetes deployments to provide better isolation.  

> Firstly, I don't think you answered my above question of effectively "How will kubectl logs work off just a logdirectory" and related details. I'm still convinced that a GetLogs method on the ContainerRuntime interface is the most sane way to support that operation, if not all logs operations (sure, something like your log collector probably should do something else, but hey, why should that be a log directory vs just knowing to forward journald logs?)

`kubectl logs` issue can be solved for example by injecting a wrapper binary between the docker daemon and the actual container process. The wrapper binary can re-direct logs to a kubelet specified file path and then exec the user process. 
Frankly, the integration plan for existing runtimes isn't clear atm. For example, we can provide a backdoor legacy logging API for docker and rkt to begin with. But, I'd prefer at least the new runtimes to follow certain best practices when it comes to logging and disk usage.


> If the eventual goal is each of this functionalities you mentioned, then why not require that the Pod/Container Config structs have a log rotation policy, a log retention policy, and a diskio limit?

Log rotation and retention will be provided on a best effort basis. The recommendation is to ship log files off the nodes to a central logging service. So I don't see a need for it to be in the API at all. 
But for Disk IO, we can wait for user feedback to see if they want to configure it. Until then, kubelet can at-least expose disk IO metrics accurately per container and let users kill offending pods.



> Expressing the desire for those features to a runtime in this contract seems like the clear way to make it have a chance of it happening. Not asking the runtimes to toss data at some directory and adding more code in the runtime to try and rotate log-format XYZ without corrupting it.

If I was not clear, apologies. This proposal is excepting runtimes to re-direct stdout & stderr to a kubelet specified file, without clobbering the log lines, at-least keeping it human readable. If you run journald as well, thats probably fine, but the disk usage needs to be accounted appropriately. 

> If you're okay with rkt dealing with logging altogether, then why care at all about the logpath in general? If one container runtime need not use it (because of the above), then it doesn't really make sense for it to be in a general interface.
 
I disagree. Design choices of one runtime should not ideally affect the entire ecosystem.